source/as-promise/index.ts

```diff
@@ -0,0 +1,242 @@
+import {EventEmitter} from 'events';
+import getStream = require('get-stream');
+import PCancelable = require('p-cancelable');
+import calculateRetryDelay from './calculate-retry-delay';
+import {
+	NormalizedOptions,
+	CancelableRequest,
+	Response,
+	RequestError,
+	HTTPError,
+	ReadError,
+	ParseError
+} from './types';
+import PromisableRequest, {parseBody} from './core';
+import proxyEvents from '../core/utils/proxy-events';
+
+export default function asPromise<T>(options: NormalizedOptions): CancelableRequest<T> {
+	let retryCount = 0;
+	let body: Buffer;
+	const emitter = new EventEmitter();
+
+	const promise = new PCancelable<T>((resolve, reject, onCancel) => {
+		const makeRequest = (): void => {
+			if (options.responseType === 'json' && options.headers.accept === undefined) {
+				options.headers.accept = 'application/json';
+			}
+
+			// Support retries
+			const {throwHttpErrors} = options;
+			if (!throwHttpErrors) {
+				options.throwHttpErrors = true;
+			}
+
+			const request = new PromisableRequest(options.url, options);
+			request._throwHttpErrors = throwHttpErrors;
+			request._noPipe = true;
+			onCancel(() => request.destroy());
+
+			request.once('response', async (response: Response) => {
+				response.retryCount = retryCount;
+
+				if (response.request.aborted) {
+					// Canceled while downloading - will throw a `CancelError` or `TimeoutError` error
+					return;
+				}
+
+				const isOk = (): boolean => {
+					const {statusCode} = response;
+					const limitStatusCode = options.followRedirect ? 299 : 399;
+
+					return (statusCode >= 200 && statusCode <= limitStatusCode) || statusCode === 304;
+				};
+
+				// Download body
+				try {
+					body = await getStream.buffer(request);
```

Comment:
Note that this is `request`, not `response`. If we were downloading directly from the `response` there would be no progress events emitted.